File: Personal Development/leetcode/insertion_sort_list.go
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func insertionSortList(head *ListNode) *ListNode {
	if head.Next == nil {
		return head
	}
	ints := make([]int, 0)
	for head != nil {
		ints = append(ints, head.Val)
		head = head.Next
	}
	sort.Ints(ints)
	ans := new(ListNode)
	ite := ans
	for i := 0; i < len(ints); i++ {
		ite.Val = ints[i]
		if i+1 < len(ints) {
			ite.Next = &ListNode{}
			ite = ite.Next
		}
	}
	return ans

}

File: A2SV/ProgressSheet/leetcode/leetcode/insertion-sort-list.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def insertionSortList(self, head: Optional[ListNode]) -> Optional[ListNode]:
        d1 = ListNode(-1, head)
        prev, curr = head, head.next
        while curr:
            if curr.val >= prev.val:
                prev, curr = curr, curr.next
                continue
            
            temp = d1
            while curr.val > temp.next.val:
                temp = temp.next
            
            prev.next = curr.next
            curr.next = temp.next
            temp.next = curr
            curr = prev.next
        
        return d1.next
